main/src/ca/uwaterloo/flix/language/phase/Uncurrier2.scala

```diff
@@ -0,0 +1,366 @@
+import ca.uwaterloo.flix.util.{InternalCompilerException, Validation}
+
+object Uncurrier2 extends Phase[SimplifiedAst.Root, SimplifiedAst.Root] {
+
+  override def run(root: SimplifiedAst.Root)(implicit flix: Flix): Validation[SimplifiedAst.Root, CompilationError] = flix.phase("Uncurrier2") {
+    /* Uncurry an arrow type A -> (B -> C) into an arrow (A, B) -> C
+     * where A may be multiple types and A, B, C are all proper types.
+    */
+    def uncurryArrow(tpe0: Type): Type = tpe0 match {
+      case Type.Apply(lhs0, Type.Apply(Type.Apply(Type.Arrow(2, eff1), b), c)) => //TODO: arity could be more than 2
+        def visitLhs(lhs0: Type): Type = lhs0 match {
+          case Type.Arrow(arity0, eff0) =>
+            val eff2 = eff1 match {
+              case Type.Impure => Type.Impure
+              case Type.Pure => eff0
+              case Type.Unit => Type.Unit //TODO: What does this mean?
+              case _ => throw InternalCompilerException(s"Unexpected effect: '${eff1.getClass}'.")
+            }
+            Type.Arrow(arity0 + 1, eff2)
+
+          case Type.Apply(lhs0, rhs0) =>
+            val lhs1 = visitLhs(lhs0)
+            Type.Apply(lhs1, rhs0)
+
+          case _ => throw InternalCompilerException(s"Unexpected type: '${lhs0.getClass}'.")
+        }
+
+        val lhs1 = visitLhs(lhs0)
+        val lhs2 = Type.Apply(lhs1, b)
+        Type.Apply(lhs2, c)
+
+      case _ => throw InternalCompilerException(s"Unexpected type: '${tpe0.getClass}'.")
+    }
+
+    def uncurryDefn(defn: (Symbol.DefnSym, Def), acc0: Map[Symbol.DefnSym, Def], ctxt0: Map[Symbol.DefnSym, Symbol.DefnSym]): (Map[Symbol.DefnSym, Def], Map[Symbol.DefnSym, Symbol.DefnSym]) = {
+      val (sym0, defn0) = defn
+      if (defn0.ann.isBenchmark || defn0.ann.isTest) {
+        (acc0, ctxt0)
+      } else {
+        defn0.exp match {
+          case Expression.Lambda(fparams, body, tpe, loc) =>
+            val sym1 = Symbol.freshDefnSym(sym0)
+            val params0 = defn0.fparams ++ fparams
+            val params1 = params0.map(fparam => fparam.copy(sym = Symbol.freshVarSym(fparam.sym)))
+            val vars0 = params0.zip(params1).map {
+              case (oldParam, newParam) => oldParam.sym -> newParam.sym
+            }
+            val exp1 = SimplifiedAstOps.renameBoundVars(body, vars0.toMap, Map.empty)
+            val tpe1 = uncurryArrow(defn0.tpe)
+            val uncurriedDefn = defn0.copy(sym = sym1, fparams = params1, exp = exp1, tpe = tpe1)
+            val result = sym1 -> uncurriedDefn
+            val acc1 = acc0 + result
+            val (us, ctxt1) = uncurryDefn(result, acc1, ctxt0)
+            val ctxt2 = ctxt1 + (defn0.sym -> sym1)
+            (us, ctxt2)
+
+          case _ => (acc0, ctxt0)
+        }
+      }
+    }
+
+    val (uncurriedDefs0, uncurryCtxt) = flix.subphase("Uncurrying step 0") {
+      root.defs.foldLeft((Map.empty[Symbol.DefnSym, Def], Map.empty[Symbol.DefnSym, Symbol.DefnSym])) {
```

Comment:
I am not sure I understand why this is a map from sym to sym without any notion of arity?
Can you write the invariant for this map?

Reply:
Given a curried top-level def foo, we generate defs foo$1, foo$2, ..., foo$n where foo$1 is the original def foo. foo$2 has arity 2, etc.

Invariant: If foo$i is a key in the map, then it is mapped to foo$i+1 and no other keys map to foo$i+1.

Or did you mean I should add that as a comment in the code?

Reply:
Yes, please it to the code.